Nodeupgrade

guineveresaenger opened the pull request:
Inviting comments on this for now. Upgrade works for AWS as far as I can tell.

Things to consider:
- does it make sense to have three separate yaml files per nodepool? I like it because although it looks like a lot of duplicated code, it keeps the concerns separate. However, I feel myself wanting to refactor what is being put into the kill_node filter. I'm not sure how to do that.
- I still would like to find a better way of upgrading than requiring an `up.sh` before runnong `upgrade.sh`. Wouldn't it be nice if we only ran necessary tasks? It would also save time. I learned today that this is because we need to update the cloudinit to use the new version. I tried including the assembler task in kraken.upgrade, but while it did update the cloud-init file, the nodes themselves didn't update their versions.
- This upgrade takes quite a long time, about ten minutes per node. That's over an hour, worse than tearing down a cluster and putting up a new one.
- Can I please delete the very confusing update.yaml? It looks like somebody's lost orphan child, so unless it has a purpose I'm not aware of...?

guineveresaenger commented:
Oh yeah. I haven't finished the GKE work on this yet.

guineveresaenger commented:
retest this please

NathanZook commented:
Yeah, I'm nervous about all of the code duplication.  Especially since we are in an exploratory mode in so many ways, this looks like trouble.

guineveresaenger commented:
(This adresses issues in #85)

coffeepac commented on ansible/roles/kraken.upgrade/kraken.upgrade.aws/tasks/main.yaml:
```diff
@@ -5,4 +5,15 @@
-  
\ No newline at end of file
+
+- include: aws-cluster-nodes.yaml
```
this is not going to work.  the names of the nodePools are arbitrary and we will want this to work going forward.  

we made a short term decision to hardcode the master nodePool name as that's a reasonable default most people won't change, but odds are high that people will add additional nodePools as we go forward.

for now I think it would suffice to either upgrade the master nodePool OR all other nodePools.  we can work out how to handle specify individual nodePools later

guineveresaenger replied:
I see. Then my question is (and this will make the files look so much less cluttered): do we really need to separate out control plane and "other" nodes? The main reason for treating control plane nodes differently was to give each node time to respawn, so the remaining nodes can keep the cluster running in the meantime. The kill_node filter takes care of that. Are there other arguments for giving users the capability to upgrade only parts of the cluster, specifically "control plane" vs "everything else"?
We don't have to worry about hard coding anything in that case.

guineveresaenger commented:
Ouch. It would be great to be able to terminate the ci check if you do something stupid like forget to rebase or accidentally left something commented out...

guineveresaenger commented:
The test failed on terraform destroy.

guineveresaenger commented:
retest this please

guineveresaenger commented:
@coffeepac  do you mind re-testing this for me?

guineveresaenger commented:
PR: #473

coffeepac commented:
I am testing this locally now.  Should be done by lunch.

guineveresaenger commented:
@coffeepac my tests show they passed?

coffeepac commented:
@guineveresaenger excellent! this is large enough that it needs someone else to operate it to make sure it works on not your machine.  if we had a better testing framework in place for this type of activity (next milestone maybe?) we would write tests to exercise this new functionality automatically.  lacking that, a second human is advised.

guineveresaenger commented:
@coffeepac that makes so much sense! and makes me feel so much better about changes that, um, don't get used on the automatic tests... hopefully the documentation rewrite is clear.

guineveresaenger commented:
@coffeepac I think I fixed the documentation to talk about the `--nodepools` flag, and I added a line to the help text that mentions `./upgrade.sh`. Please let me know if that takes care of the explanations part!

guineveresaenger commented:
@coffeepac this did fail e2e tests as far as I can tell. I'll trigger it one more time. One more dayyyyyyy

guineveresaenger commented:
As per comments in #85, this is not yet ready to merge. Apologies.

guineveresaenger commented:
unless we want to merge this, knowing it will work for AWS but not gke.

guineveresaenger commented:
Should be ready to go now.
